File: app/src/main/java/com/aavens/mindloft/ui/main/MainActivity.java
package com.aavens.mindloft.ui.main;

import android.content.Intent;
import android.os.Bundle;
import android.support.design.widget.FloatingActionButton;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.view.View;
import android.view.Menu;
import android.view.MenuItem;

import com.aavens.mindloft.R;
import com.aavens.mindloft.listeners.DialogFragmentActionsListener;
import com.aavens.mindloft.listeners.OnRoomClickListener;
import com.aavens.mindloft.managers.RoomsManager;
import com.aavens.mindloft.models.Room;
import com.aavens.mindloft.ui.detail.RoomActivity;
import java.util.Observable;
import java.util.Observer;

public class MainActivity
        extends AppCompatActivity
        implements
        OnRoomClickListener,
        Observer,
        DialogFragmentActionsListener {

    private RecyclerView recyclerView;
    private RecyclerView.Adapter adapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        RoomsManager.getInstance().setContext(this);
        RoomsManager.getInstance().addObserver(this);
        initToolbar();
        initFloatingButton();
        initRecyclerView();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_main, menu);
        return true;
    }

    // MARK: OnRoomClickListener

    @Override
    public void onItemClick(Integer index) {
        navigateToRoomActivity(index);
    }

    @Override
    public void onItemLongClick(Integer index) {
        RoomActionsDialogFragment fragment = RoomActionsDialogFragment.newInstance(index, this);
        fragment.show(getFragmentManager(), "RoomActionsDialogFragment");
    }

    // MARK: Observer

    @Override
    public void update(Observable observable, Object data) {
        RoomsManager.ActionInfo actionInfo = (RoomsManager.ActionInfo) data;
        int position = actionInfo.getPosition();
        switch (actionInfo.getAction()) {
            case ADD:
                adapter.notifyItemInserted(position);
                recyclerView.scrollToPosition(position);
                break;
            case UPDATE:
                adapter.notifyItemChanged(position);
                break;
            case REMOVE:
                adapter.notifyItemRemoved(position);
                break;
        }
    }

    // MARK: DialogFragmentActionsListener

    public void onActionInteraction(Integer index) {
        Room room = RoomsManager.getInstance().get(index);
        RoomsManager.getInstance().delete(room);
    }

    // MARK: Private

    private void initToolbar() {
        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
    }

    private void initFloatingButton() {
        FloatingActionButton fab = (FloatingActionButton) findViewById(R.id.fab);
        if (fab == null) {
            return;
        }
        fab.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                navigateToRoomActivity(null);
            }
        });
    }

    private void initRecyclerView() {
        recyclerView = (RecyclerView) findViewById(R.id.rooms_list_view);
        adapter = new RoomsRecyclerViewAdapter(this);
        recyclerView.setAdapter(adapter);
        LinearLayoutManager linearLayoutManager = new LinearLayoutManager(this);
        linearLayoutManager.setOrientation(LinearLayoutManager.VERTICAL);
        recyclerView.setLayoutManager(linearLayoutManager);
    }

    private void navigateToRoomActivity(Integer index) {
        Intent intent = new Intent(this, RoomActivity.class);
        intent.putExtra(RoomActivity.EXTRA_ROOM_INDEX, index);
        startActivity(intent);
    }
}

File: app/src/main/java/com/aavens/mindloft/ui/main/MainActivity.kt
package com.aavens.mindloft.ui.main

import android.content.Intent
import android.os.Bundle
import android.support.design.widget.FloatingActionButton
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.support.v7.widget.Toolbar
import android.view.Menu
import com.aavens.mindloft.R
import com.aavens.mindloft.listeners.DialogFragmentActionsListener
import com.aavens.mindloft.listeners.OnRoomClickListener
import com.aavens.mindloft.managers.RoomsManager
import com.aavens.mindloft.ui.detail.RoomActivity
import java.util.*

class MainActivity: AppCompatActivity(),
        Observer,
        OnRoomClickListener,
        DialogFragmentActionsListener {

    private var recyclerView: RecyclerView? = null
    private var adapter: RecyclerView.Adapter<RoomRecycleViewAdapter.ViewHolder>? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        RoomsManager.instance.context = this
        RoomsManager.instance.addObserver(this)
        initToolbar()
        initFloatingButton()
        initRecyclerView()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_main, menu)
        return true
    }

    // MARK: OnRoomClickListener

    override fun onItemClick(index: Int?) {
        navigateToRoomActivity(index)
    }

    override fun onItemLongClick(index: Int?) {
        if (index != null) {
            val fragment = RoomActionsDialogFragment(index.toInt(), this)
            fragment.show(fragmentManager, "RoomActionsDialogFragment")
        }
    }

    // MARK: Observer

    override fun update(observable: Observable, data: Any) {
        val actionInfo = data as RoomsManager.ActionInfo
        val position = actionInfo.position
        when (actionInfo.action) {
            RoomsManager.Action.ADD -> {
                adapter?.notifyItemInserted(position)
                recyclerView?.scrollToPosition(position)
            }
            RoomsManager.Action.UPDATE -> adapter?.notifyItemChanged(position)
            RoomsManager.Action.REMOVE -> adapter?.notifyItemRemoved(position)
        }
    }

    // MARK: DialogFragmentActionsListener

    override fun onActionInteraction(index: Int?) {
        if (index != null) {
            val room = RoomsManager.instance[index]
            RoomsManager.instance.delete(room)
        }
    }

    // MARK: Private

    private fun initToolbar() {
        val toolbar = findViewById(R.id.toolbar) as Toolbar?
        setSupportActionBar(toolbar)
    }

    private fun initFloatingButton() {
        val fab = findViewById(R.id.fab) as FloatingActionButton? ?: return
        fab.setOnClickListener { navigateToRoomActivity(null) }
    }

    private fun initRecyclerView() {
        recyclerView = findViewById(R.id.rooms_list_view) as RecyclerView?
        adapter = RoomRecycleViewAdapter(this)
        recyclerView?.adapter = adapter
        val linearLayoutManager = LinearLayoutManager(this)
        linearLayoutManager.orientation = LinearLayoutManager.VERTICAL
        recyclerView?.layoutManager = linearLayoutManager
    }

    private fun navigateToRoomActivity(index: Int?) {
        val intent = Intent(this, RoomActivity::class.java)
        intent.putExtra(RoomActivity.EXTRA_ROOM_INDEX, index)
        startActivity(intent)
    }
}
